File: EVSDKDemo/EVSDKDemo/EVDemoWatchViewController.m
//
//  EVDemoWatchViewController.m
//  EVSDKDemo
//
//  Created by mashuaiwei on 16/8/12.
//  Copyright © 2016年 cloudfocous. All rights reserved.
//

#import "EVDemoWatchViewController.h"
#import "EVPlayer.h"
#import "EVPlayerConfig.h"
#import "EVStreamer.h"
#import "CCAlertManager.h"

@interface EVDemoWatchViewController ()<EVPlayerDelegate, EVStreamerDelegate>

@property (weak, nonatomic) IBOutlet UIView *containerView;
@property (weak, nonatomic) IBOutlet UIButton *closeBtn;
@property (nonatomic, strong) EVPlayer *player; /**< 播放器 */
@property (nonatomic, strong) EVStreamer *streamer; /**< 推流端 */
@property (nonatomic, copy) NSString *agoraAppId;

@end

@implementation EVDemoWatchViewController

#pragma mark - life circle

- (void)viewDidLoad{
    [super viewDidLoad];
    
//    self.agoraAppId = @"your agoraId";
    [self setUpPlayer];
}


#pragma mark - UIAction

- (IBAction)close:(UIButton *)sender {
    [self.player shutDown];
    _player = nil;
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (IBAction)interactiveLiveButtonClicked:(UIButton *)sender {
    if (!self.agoraAppId) {
        [[CCAlertManager shareInstance] performComfirmTitle:nil message:@"请先设置agoraAppid" comfirmTitle:@"确定" WithComfirm:nil];
        return;
    }
    sender.selected = !sender.selected;
    if (sender.selected) {
        [self.player pause];
        if (!self.streamer) {
            self.streamer = [[EVStreamer alloc] init];
            self.streamer.delegate = self;
            self.streamer.frontCamera = YES;
            [self.containerView addSubview:self.streamer.preview];
            self.streamer.preview.frame = self.containerView.bounds;
            self.streamer.agoraAppid = self.agoraAppId;
            [self.streamer livePrepareComplete:nil];
            [self.streamer startPreview];
        }
        [self.streamer joinChannel:@"evdemo02"];
    } else {
        [self.streamer leaveChannel];
    }
}


#pragma mark - EVStreamerDelegate
- (void)EVStreamerUpdateVideoChatState:(EVVideoChatState)chatState{
    NSLog(@"连麦状态：%lu", chatState);
    if (chatState == EVVideoChatStateChannelLeaved) {
        [self.streamer stopPreview];
        [self.streamer.preview removeFromSuperview];
        _streamer = nil;
        [self.player play];
    }
}


#pragma mark - private methods

- (void)setUpPlayer{
    self.player = [[EVPlayer alloc] init];
    self.player.playerContainerView = self.containerView;
    self.player.live = YES;
    self.player.lid = self.vid;
    
    __weak typeof(self) wSelf = self;
    [self.player playPrepareComplete:^(EVPlayerResponseCode responseCode, NSDictionary *result, NSError *err) {
        __strong typeof(wSelf) sSelf = wSelf;
        if (responseCode == EVPlayerResponse_Okay) {
            [sSelf.player play];
        }
    }];
}

@end

File: EVSDKDemo_Swift/Player/EVDemoWatchViewController.swift

//
//  EVDemoWatchViewController.swift
//  EVSDKDemo
//
//  Created by Lcrnice on 2017/8/10.
//  Copyright © 2017年 cloudfocous. All rights reserved.
//

import UIKit
import EVMediaFramework

class EVDemoWatchViewController: UIViewController {
  
  var lid = ""
  fileprivate var player: EVPlayer?
  fileprivate var streamer: EVStreamer?
  private let kAgoraId = ""
		
  @IBOutlet weak var closeBtn: UIButton!
  @IBOutlet weak var containerView: UIView!
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    self.setupPlayer()
  }
  
  override func didReceiveMemoryWarning() {
    super.didReceiveMemoryWarning()
  }
  
  @IBAction func close(_ sender: Any) {
    if streamer?.chatState == .remoteVideoRendered {
      streamer?.leaveChannel()
    } else {
      player?.shutDown()
    }
    
    self.dismiss(animated: true, completion: nil)
  }
  
  @IBAction func interactiveLive(_ sender: UIButton) {
    sender.isSelected = !sender.isSelected
    
    if sender.isSelected {
      player?.pause()
      
      if streamer == nil {
        streamer = EVStreamer()
        streamer?.delegate = self as EVStreamerDelegate
        streamer?.frontCamera = true
        containerView.addSubview((streamer?.preview!)!)
        streamer?.preview.frame = containerView.frame
        streamer?.agoraAppid = kAgoraId
        streamer?.livePrepareComplete(nil)
        streamer?.startPreview()
      }
      
      if (streamer?.agoraAppid.isEmpty)! {
        CCAlertManager.shareInstance().performComfirmTitle("提示", message: "请先设置agoraAppid", comfirmTitle: "OK", withComfirm: nil)
        return
      }
      
      streamer?.joinChannel("evtest002")
    } else {
      streamer?.leaveChannel()
    }
  }
  
  private func setupPlayer() {
    player = EVPlayer()
    player?.playerContainerView = containerView
    player?.live = true
    player?.lid = lid
    
    player?.playPrepareComplete { [weak self] (code: EVPlayerResponseCode, info: [AnyHashable : Any]?, error: Error?) in
      guard let sSelf = self else {
        return
      }
      
      if code == .okay {
        sSelf.player?.play()
      } else {
        print("获取直播地址失败：\(String(describing: error?.localizedDescription))")
      }
    }
  }
  
  
}


extension EVDemoWatchViewController: EVStreamerDelegate {
  func evStreamerUpdate(_ chatState: EVVideoChatState) {
    if chatState == .channelLeaved {
      streamer?.stopPreview()
      streamer?.preview.removeFromSuperview()
      streamer = nil
      player?.play()
    }
  }
}
